### DISCUSSION TITLE
Using rust types with flutter rust bridge

### DISCUSSION THREAD

**Discussion by Likho** [fzyzcjy/flutter_rust_bridge#420] [Q&A]
Hi I am new to rust and flutter. I have a function that returns the type below 
```
type Wallet = Arc<
    Mutex<
        Box<
            dyn WalletInst<
                'static,
                DefaultLCProvider<'static, HTTPNodeClient, ExtKeychain>,
                HTTPNodeClient,
                ExtKeychain,
            >,
        >,
    >,
>;
```
Running flutter_rust_bridge_codegen for my fuction returns 'Failed to parse function output type `Result<Wallet, Error>`. How do I handle rust types in flutter_rust_bridge. Thanks

**Comment by fzyzcjy** [comment]
Well, `Arc` and `Mutex` and maybe `dyn` are not supported - you know, they cannot be serialized. What about creating custom structs and pass them around?

**Comment by Desdaemon** [comment]
Smart pointers are not supported yet (they are part of #293), but if you must use them you can create a raw pointer from them and use a newtype struct to wrap it.

**Reply by Likho** [reply]
Thank you, I will read more on this. The documentation does not get into much detail about importing pub structs with private fields, is this supported yet?

**Reply by Desdaemon** [reply]
For now all exported Rust structs must only have public members, but I'm working on something to relax this restriction.
